crates/pico/src/memo.rs

```diff
@@ -0,0 +1,189 @@
+use pico_core::{
+    container::Container,
+    database::Database,
+    dyn_eq::DynEq,
+    epoch::Epoch,
+    key::Key,
+    node::{Dependency, DerivedNode, DerivedNodeId, NodeKind},
+    params::ParamId,
+    storage::{Storage, StorageMut},
+};
+
+pub enum MemoState {
+    Memoized,
+    Computed,
+}
+
+pub fn memo<Db: Database>(
+    db: &mut Db,
+    node_id: DerivedNodeId,
+    inner_fn: fn(&mut Db, ParamId) -> Box<dyn DynEq>,
+) -> MemoState {
+    let current_epoch = db.current_epoch();
+    let (time_updated, state) = if db.storage().derived_nodes().contains_key(&node_id) {
```

Comment:
sorry if I've asked this before, but is this more idiomatically done via `match derived_nodes.entry(...)`? In particular, that might allow us to avoid the `expect` in `any_dependency_changed`.

Maybe it's the fact that we're calling `call_inner_fn` etc that prevents us from keeping a mutable reference to the item around...

Reply:
yeah, the problem is we can't hold reference to node and use `&mut db` for `call_inner_fn` the same time